rx-java/src/test/java/io/vertx/rx/java/test/WorkerExecutorSchedulerTest.java

```diff
@@ -0,0 +1,254 @@
+    Observable<String> observable = Observable.<String>create(subscriber -> {
+      assertFalse(Context.isOnVertxThread());
+      subscriber.onNext("expected");
+      subscriber.onCompleted();
+    }).delay(10, TimeUnit.MILLISECONDS, scheduler).doOnNext(o -> assertWorkerExecutorThread());
+    new Thread(() -> {
+      observable.subscribe(
+        item -> assertEquals("expected", item),
+        this::fail,
+        this::testComplete);
+    }).start();
+    await();
+  }
+
+  @Test
+  public void testScheduleDelayed() {
+    Scheduler.Worker worker = scheduler.createWorker();
+    long time = System.currentTimeMillis();
+    worker.schedule(() -> {
+      assertWorkerExecutorThread();
+      assertTrue(System.currentTimeMillis() - time >= 40);
+      testComplete();
+    }, 40, TimeUnit.MILLISECONDS);
+    await();
+  }
+
+  @Test
+  public void testSchedulePeriodic() {
+    Scheduler.Worker worker = scheduler.createWorker();
+    AtomicLong time = new AtomicLong(System.currentTimeMillis() - 40);
+    AtomicInteger count = new AtomicInteger();
+    AtomicReference<Subscription> sub = new AtomicReference<>();
+    sub.set(worker.schedulePeriodically(() -> {
+      assertWorkerExecutorThread();
+      if (count.incrementAndGet() > 2) {
+        sub.get().unsubscribe();
+        testComplete();
+      } else {
+        long now = System.currentTimeMillis();
+        long delta = now - time.get();
+        assertTrue("" + delta, delta >= 40);
+        time.set(now);
+      }
+    }, 0, 40, TimeUnit.MILLISECONDS));
+    await();
+  }
+
+  @Test
+  public void testUnsubscribeBeforeExecute() throws Exception {
+    Scheduler.Worker worker = scheduler.createWorker();
+    CountDownLatch latch = new CountDownLatch(1);
+    Subscription sub = worker.schedule(latch::countDown, 20, TimeUnit.MILLISECONDS);
+    sub.unsubscribe();
+    assertFalse(latch.await(40, TimeUnit.MILLISECONDS));
+  }
+
+  @Test
+  public void testUnsubscribeDuringExecute() throws Exception {
+    Scheduler.Worker worker = scheduler.createWorker();
+    AtomicInteger count = new AtomicInteger();
+    AtomicReference<Subscription> sub = new AtomicReference<>();
+    sub.set(worker.schedulePeriodically(() -> {
+      if (count.getAndIncrement() == 0) {
+        sub.get().unsubscribe();
+      }
+    }, 0, 5, TimeUnit.MILLISECONDS));
+    Thread.sleep(60);
+    assertEquals(1, count.get());
+  }
+
+  @Test
+  public void testUnsubscribeBetweenActions() throws Exception {
+    Scheduler.Worker worker = scheduler.createWorker();
+    AtomicInteger count = new AtomicInteger();
+    CountDownLatch latch = new CountDownLatch(1);
+    AtomicReference<Subscription> sub = new AtomicReference<>();
+    sub.set(worker.schedulePeriodically(() -> {
+      if (count.incrementAndGet() == 4) {
+        latch.countDown();
+      }
+    }, 0, 20, TimeUnit.MILLISECONDS));
+    awaitLatch(latch);
+    sub.get().unsubscribe();
+    Thread.sleep(60);
```

Comment:
The latch helps to wait until 4 actions have been scheduled. The sleep call
is to make sure that, after a period arbitrary longer than specified in the
schedulePeriodically call, actions were not scheduled any more.

2016-08-11 1:37 GMT+02:00 testn <notifications@github.com>:

> In rx-java/src/test/java/io/vertx/rx/java/test/
> WorkerExecutorSchedulerTest.java
> <https://github.com/vert-x3/vertx-rx/pull/46#discussion_r74350926>:
>
> > +  }
> > +
> > +  @Test
> > +  public void testUnsubscribeBetweenActions() throws Exception {
> > +    Scheduler.Worker worker = scheduler.createWorker();
> > +    AtomicInteger count = new AtomicInteger();
> > +    CountDownLatch latch = new CountDownLatch(1);
> > +    AtomicReference<Subscription> sub = new AtomicReference<>();
> > +    sub.set(worker.schedulePeriodically(() -> {
> > +      if (count.incrementAndGet() == 4) {
> > +        latch.countDown();
> > +      }
> > +    }, 0, 20, TimeUnit.MILLISECONDS));
> > +    awaitLatch(latch);
> > +    sub.get().unsubscribe();
> > +    Thread.sleep(60);
>
> why sleep need to be added here if you already have a latch?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/vert-x3/vertx-rx/pull/46/files/e7488c91f845d8c98eaad4b9bb16839546e7b18a#r74350926>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABbltiAsPXK4GYIRRMfiG7aL5lRLOOT5ks5qemC6gaJpZM4JhJfH>
> .
>